src/liboslexec/batched_analysis.cpp

```diff
@@ -0,0 +1,2417 @@
+static ustring op_endswith("endswith");
+static ustring op_eq("eq");
+static ustring op_functioncall("functioncall");
+static ustring op_functioncall_nr("functioncall_nr");
+static ustring op_ge("ge");
+static ustring op_getattribute("getattribute");
+static ustring op_getchar("getchar");
+static ustring op_getmatrix("getmatrix");
+static ustring op_getmessage("getmessage");
+static ustring op_gt("gt");
+static ustring op_hash("hash");
+static ustring op_if("if");
+static ustring op_le("le");
+static ustring op_lt("lt");
+static ustring op_neq("neq");
+static ustring op_or("or");
+static ustring op_pow("pow");
+static ustring op_return("return");
+static ustring op_startswith("startswith");
+static ustring op_stoi("stoi");
+static ustring op_stof("stof");
+static ustring op_strlen("strlen");
+static ustring op_substr("substr");
+static ustring op_surfacearea("surfacearea");
+static ustring op_trace("trace");
+static ustring op_transform("transform");
+static ustring op_transformv("transformv");
+static ustring op_transformn("transformn");
+static ustring op_texture("texture");
+static ustring op_texture3d("texture3d");
+
+// Shader global strings
+static ustring object2common("object2common");
+static ustring shader2common("shader2common");
+static ustring flipHandedness("flipHandedness");
+}  // namespace Strings
+
+namespace pvt {
+
+// Implemented in batched_backendllvm.cpp
+extern bool
+is_shader_global_uniform_by_name(ustring name);
+
+
+// When adding new ops, or when changing implementation or behavior of
+// existing ops, the following functions should be evaluated to see if
+// they need to be updated to handle to handle the new op or change:
+//     bool are_op_results_always_implicitly_varying(ustring opname);
+//     bool are_op_results_always_implicitly_uniform(ustring opname);
+//     bool does_op_implementation_require_masking(ustring opname);
+//     SymbolPtrVec * protected_shader_globals_op_implicitly_depends_on(ShaderInstance &inst, Opcode & opcode);
+// NOTE: psg stands for ProtectedShaderGlobals which are explained below...
+
+// If there is a new or change in control flow op's are made,
+// a more detailed evaluation will be needed as their behavior
+// is hard coded below vs. lookup based implementation of the
+// functions above.
+
+namespace  // Unnamed
+{
+// Even when all inputs to an operation are uniform,
+// the results written to output symbols maybe varying.
+bool
+are_op_results_always_implicitly_varying(ustring opname)
+{
+    // Test if explicit comparison is faster or not
+    static std::unordered_set<ustring, ustringHash> lazy_lookup(
+        { // Renderer might identify result of getattribute as always uniform
+          // depending on the attribute itself, so will have a special
+          // handler for it during discovery
+          Strings::op_getmessage, Strings::op_trace, Strings::op_texture,
+          Strings::op_texture3d });
+    return lazy_lookup.find(opname) != lazy_lookup.end();
+}
```

Comment:
As these are 1 time population, then flat_set/flat_map would be a good choice.
Is 4 ustring comparisons faster than a flat_set lookup, maybe (probably).

Reply:
Good question, I don't know. 

For a write-once map, generally the flat_map can substantially outperform unordered_map for lookup time at nearly any size, even though lookups are O(log n) vs O(1), respectively, because the flat_map has its values stored contiguously which is great for cache locality and doesn't have extra indirections for each access.

I imagine that, as you imply, for some sufficiently small size, O(n) probe of every element is better than O(log n) with the extra logic of the binary search. But I wouldn't dare to guess without a benchmark. If it's truly 3 or 4 entries, frankly I would just do a linear probe on an array and not even bother with flat_map, but that's just my intuition.